java.nio.file.NoSuchFileException: flank/flank/CostReport.txt

I'm getting this error when running flank 4.4.0:

```
[...]
FetchArtifacts
  ...............................................................................................
  Updating matrix file

CostReport
  Virtual devices
    $0.33 for 20m

Exception in thread "main" picocli.CommandLine$ExecutionException: Error while running command (ftl.cli.firebase.test.android.AndroidRunCommand@7c9b78e3): java.nio.file.NoSuchFileException: flank/flank/CostReport.txt
        at picocli.CommandLine.execute(CommandLine.java:1168)
        at picocli.CommandLine.access$800(CommandLine.java:141)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:1367)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:1335)
        at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:1243)
        at picocli.CommandLine.parseWithHandlers(CommandLine.java:1526)
        at picocli.CommandLine.run(CommandLine.java:1974)
        at picocli.CommandLine.run(CommandLine.java:1920)
        at ftl.Main$Companion.main(Main.kt:41)
        at ftl.Main.main(Main.kt)
Caused by: java.nio.file.NoSuchFileException: flank/flank/CostReport.txt
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
        at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
        at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
        at java.nio.file.Files.newOutputStream(Files.java:216)
        at java.nio.file.Files.write(Files.java:3292)
        at ftl.util.Utils.write(Utils.kt:26)
        at ftl.reports.CostReport.write(CostReport.kt:42)
        at ftl.reports.CostReport.run(CostReport.kt:48)
        at ftl.reports.util.ReportManager.generate(ReportManager.kt:99)
        at ftl.run.TestRunner.newRun(TestRunner.kt:355)
        at ftl.run.TestRunner$newRun$1.invokeSuspend(TestRunner.kt)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:32)
        at kotlinx.coroutines.ResumeModeKt.resumeUninterceptedMode(ResumeMode.kt:45)
        at kotlinx.coroutines.internal.ScopeCoroutine.onCompletionInternal$kotlinx_coroutines_core(Scopes.kt:28)
        at kotlinx.coroutines.JobSupport.completeStateFinalization(JobSupport.kt:305)
        at kotlinx.coroutines.JobSupport.tryFinalizeFinishingState(JobSupport.kt:230)
        at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:799)
        at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:742)
        at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:117)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:45)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.kt:116)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:76)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:53)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:35)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
        at ftl.cli.firebase.test.android.AndroidRunCommand.run(AndroidRunCommand.kt:34)
        at picocli.CommandLine.execute(CommandLine.java:1160)
        ... 9 more
```

This is how I'm starting flank:
```
java -jar flank.jar firebase test android run
```

This is the yaml config:
```yaml
# gcloud args match the official gcloud cli
# See the docs for full gcloud details https://cloud.google.com/sdk/gcloud/reference/firebase/test/android/run
gcloud:
  # -- GcloudYml --

  ## The name of a Google Cloud Storage bucket where raw test results will be stored
  # results-bucket: tmp_flank

  ## The name of a unique Google Cloud Storage object within the results bucket where raw test results will be stored
  ## (default: a timestamp with a random suffix).
  # results-dir: tmp

  ## Enable video recording during the test. Enabled by default, use --no-record-video to disable.
  # record-video: true

  ## The max time this test execution can run before it is cancelled (default: 15m).
  ## It does not include any time necessary to prepare and clean up the target device.
  ## The maximum possible testing time is 30m on physical devices and 60m on virtual devices.
  ## The TIMEOUT units can be h, m, or s. If no unit is given, seconds are assumed.
  # timeout: 30m

  ## Invoke a test asynchronously without waiting for test results.
  # async: false

  ## The history name for your test results (an arbitrary string label; default: the application's label from the APK manifest).
  ## All tests which use the same history name will have their results grouped together in the Firebase console in a time-ordered test history list.
  # results-history-name: android-history

  ## Experimental!
  ## The number of times a TestExecution should be re-attempted if one or more\nof its test cases fail for any reason.
  ## The maximum number of reruns allowed is 10. Default is 0, which implies no reruns.
  # flaky-test-attempts: 0

  # -- AndroidGcloudYml --

  ## The path to the application binary file.
  ## The path may be in the local filesystem or in Google Cloud Storage using gs:// notation.
  ## Android App Bundles are specified as .aab, all other files are assumed to be APKs.
  app: COMPANY_global_italy_api19_debug_19.03.10270-680045955-f1d1ae12f1.apk

  ## The path to the binary file containing instrumentation tests.
  ## The given path may be in the local filesystem or in Google Cloud Storage using a URL beginning with gs://.
  test: app-global-italy-debug-androidTest.apk

  ## Automatically log into the test device using a preconfigured Google account before beginning the test.
  ## Enabled by default, use --no-auto-google-login to disable.
  # auto-google-login: true

  ## Whether each test runs in its own Instrumentation instance with the Android Test Orchestrator
  ## (default: Orchestrator is used). Disable with --no-use-orchestrator.
  ## See https://developer.android.com/training/testing/junit-runner.html#using-android-test-orchestrator
  # use-orchestrator: true

  ## A comma-separated, key=value map of environment variables and their desired values. This flag is repeatable.
  ## The environment variables are mirrored as extra options to the am instrument -e KEY1 VALUE1 … command and
  ## passed to your test runner (typically AndroidJUnitRunner)
  environment-variables:
    #  coverage: true
    #  coverageFilePath: /sdcard/
    clearPackageData: true

  ## A list of paths that will be copied from the device's storage to the designated results bucket after the test
  ## is complete. These must be absolute paths under /sdcard or /data/local/tmp
  directories-to-pull:
  - /sdcard/

  ## Monitor and record performance metrics: CPU, memory, network usage, and FPS (game-loop only).
  ## Enabled by default, use --no-performance-metrics to disable.
  # performance-metrics: true

  ## A list of one or more test target filters to apply (default: run all test targets).
  ## Each target filter must be fully qualified with the package name, class name, or test annotation desired.
  ## Supported test filters by am instrument -e … include:
  ## class, notClass, size, annotation, notAnnotation, package, notPackage, testFile, notTestFile
  ## See https://developer.android.com/reference/android/support/test/runner/AndroidJUnitRunner for more information.
  # test-targets:
  #  - class com.example.app.ExampleUiTest#testPasses

  ## A list of DIMENSION=VALUE pairs which specify a target device to test against.
  ## This flag may be repeated to specify multiple devices.
  ## The four device dimensions are: model, version, locale, and orientation.
  # device:
  # - model: NexusLowRes
  #   version: 28
  #   locale: en
  #   orientation: portrait
  # - model: NexusLowRes
  #   version: 27

flank:
  # -- FlankYml --

  ## test shards - the amount of groups to split the test suite into
  ## set to -1 to use one shard per test. default: 1
  max-test-shards: 20

  ## shard time - the amount of time tests within a shard should take
  ## when set to > 0, the shard count is dynamically set based on time up to the maxmimum limit defined by max-test-shards
  ## 2 minutes (120) is recommended.
  ## default: -1 (unlimited)
  shard-time: 2

  ## repeat tests - the amount of times to run the tests.
  ## 1 runs the tests once. 10 runs all the tests 10x
  repeat-tests: 1

  ## Google cloud storage path to store the JUnit XML results from the last run.
  smart-flank-gcs-path: gs://tmp_flank/flank/test_app_android.xml

  ## Disables sharding. Useful for parameterized tests.
  # disable-sharding: false

  ## always run - these tests are inserted at the beginning of every shard
  ## useful if you need to grant permissions or login before other tests run
  # test-targets-always-run:
  #   - class com.example.app.ExampleUiTest#testPasses

  ## regex is matched against bucket paths, for example: 2019-01-09_00:13:06.106000_YCKl/shard_0/NexusLowRes-28-en-portrait/bugreport.txt
  # files-to-download:
  #   - .*\.mp4$

  ## The billing enabled Google Cloud Platform project name to use
  project: testlab-xxxxx

  ## Local folder to store the test result. Folder is DELETED before each run to ensure only artifacts from the new run are saved.
  local-result-dir: flank
```

--- a/test_runner/src/main/kotlin/ftl/args/IArgs.kt
+++ b/test_runner/src/main/kotlin/ftl/args/IArgs.kt
@@ -1,5 +1,7 @@
 package ftl.args
 
+import ftl.args.yml.FlankYmlParams
+
 interface IArgs {
     // original YAML data
     val data: String
@@ -22,8 +24,10 @@ interface IArgs {
     val testTargetsAlwaysRun: List<String>
     val filesToDownload: List<String>
     val disableSharding: Boolean
-    val localResultDir: String?
+    val localResultDir: String
 
     // computed property
     val testShardChunks: List<List<String>>
+
+    fun useLocalResultDir() = localResultDir != FlankYmlParams.defaultLocalResultDir
 }

--- a/test_runner/src/main/kotlin/ftl/run/GenericTestRunner.kt
+++ b/test_runner/src/main/kotlin/ftl/run/GenericTestRunner.kt
@@ -2,7 +2,6 @@ package ftl.run
 
 import com.google.api.services.testing.model.TestMatrix
 import ftl.args.IArgs
-import ftl.args.yml.FlankYmlParams
 import ftl.config.FtlConstants
 import ftl.config.FtlConstants.useMock
 import ftl.json.MatrixMap
@@ -17,23 +16,22 @@ object GenericTestRunner {
         val stopwatch = StopWatch().start()
         TestRunner.assertMockUrl()
 
-        val resultsDir = if (args.localResultDir != FlankYmlParams.defaultLocalResultDir) {
-            // Only one result is stored when using --local-result-dir
-            // Delete any old results if they exist before storing new ones.
-            File(args.localResultDir).deleteRecursively()
-            "../${args.localResultDir}"
-        } else {
-            args.resultsDir ?: Utils.uniqueObjectName()
-        }
+        val runGcsPath = args.resultsDir ?: Utils.uniqueObjectName()
 
         // Avoid spamming the results/ dir with temporary files from running the test suite.
         if (useMock) {
             Runtime.getRuntime().addShutdownHook(Thread {
-                File(args.localResultDir, resultsDir).deleteRecursively()
+                File(args.localResultDir, runGcsPath).deleteRecursively()
             })
         }
 
-        return stopwatch to resultsDir
+        if (args.useLocalResultDir()) {
+            // Only one result is stored when using --local-result-dir
+            // Delete any old results if they exist before storing new ones.
+            File(args.localResultDir).deleteRecursively()
+        }
+
+        return stopwatch to runGcsPath
     }
 
     fun afterRunTests(
@@ -51,8 +49,7 @@ object GenericTestRunner {
 
         val matrixMap = MatrixMap(savedMatrices, runGcsPath)
         TestRunner.updateMatrixFile(matrixMap, config)
-
-        TestRunner.saveConfigFile(matrixMap.runPath, config)
+        TestRunner.saveConfigFile(matrixMap, config)
 
         println(FtlConstants.indent + "${savedMatrices.size} matrix ids created in ${stopwatch.check()}")
         val gcsBucket = "https://console.developers.google.com/storage/browser/" +

--- a/test_runner/src/main/kotlin/ftl/run/TestRunner.kt
+++ b/test_runner/src/main/kotlin/ftl/run/TestRunner.kt
@@ -55,15 +55,23 @@ object TestRunner {
     }
 
     fun updateMatrixFile(matrixMap: MatrixMap, args: IArgs): Path {
-        val matrixIdsPath = Paths.get(args.localResultDir, matrixMap.runPath, FtlConstants.matrixIdsFile)
+        val matrixIdsPath = if (args.useLocalResultDir()) {
+            Paths.get(args.localResultDir, FtlConstants.matrixIdsFile)
+        } else {
+            Paths.get(args.localResultDir, matrixMap.runPath, FtlConstants.matrixIdsFile)
+        }
         matrixIdsPath.parent.toFile().mkdirs()
         Files.write(matrixIdsPath, gson.toJson(matrixMap.map).toByteArray())
         return matrixIdsPath
     }
 
-    fun saveConfigFile(matrixRunPath: String, args: IArgs): Path? {
-        val configFilePath =
-            Paths.get(args.localResultDir, matrixRunPath, FtlConstants.configFileName(args))
+    fun saveConfigFile(matrixMap: MatrixMap, args: IArgs): Path? {
+        val configFilePath = if (args.useLocalResultDir()) {
+            Paths.get(args.localResultDir, FtlConstants.configFileName(args))
+        } else {
+            Paths.get(args.localResultDir, matrixMap.runPath, FtlConstants.configFileName(args))
+        }
+
         configFilePath.parent.toFile().mkdirs()
         Files.write(configFilePath, args.data.toByteArray())
         return configFilePath
@@ -209,7 +217,11 @@ object TestRunner {
                     result.iterateAll().forEach { blob ->
                         val blobPath = blob.blobId.name
                         if (artifactsList.any { blobPath.matches(it) }) {
-                            val downloadFile = Paths.get(args.localResultDir, blobPath)
+                            val downloadFile = if (args.useLocalResultDir()) {
+                                Paths.get(args.localResultDir, blobPath.substringAfter("/"))
+                            } else {
+                                Paths.get(args.localResultDir, blobPath)
+                            }
                             print(".")
                             if (!downloadFile.toFile().exists()) {
                                 downloadFile.parent.toFile().mkdirs()

--- a/test_runner/src/main/kotlin/ftl/util/MatrixUtil.kt
+++ b/test_runner/src/main/kotlin/ftl/util/MatrixUtil.kt
@@ -7,6 +7,8 @@ import java.nio.file.Paths
 import java.util.concurrent.TimeUnit
 
 fun resolveLocalRunPath(matrices: MatrixMap, args: IArgs): String {
+    if (args.useLocalResultDir()) return args.localResultDir
+
     var runPath = File(matrices.runPath)
     if (!runPath.exists()) runPath = Paths.get(args.localResultDir, runPath.name).toFile()
 

